src/core/lib/surface/completion_queue.cc

```diff
@@ -821,15 +817,19 @@ static void cq_end_op_for_pluck(grpc_completion_queue* cq, void* tag,
   GRPC_ERROR_UNREF(error);
 }
 
+static void functor_callback(void* arg, grpc_error* error) {
+  auto* functor = static_cast<grpc_experimental_completion_queue_functor*>(arg);
+  functor->functor_run(functor, error == GRPC_ERROR_NONE);
+}
+
 /* Complete an event on a completion queue of type GRPC_CQ_CALLBACK */
 static void cq_end_op_for_callback(
     grpc_completion_queue* cq, void* tag, grpc_error* error,
     void (*done)(void* done_arg, grpc_cq_completion* storage), void* done_arg,
-    grpc_cq_completion* storage) {
+    grpc_cq_completion* storage, bool internal) {
```

Comment:
I think I may have completely mislead you on the internal thing. The thing is that the CQ tag for incoming server calls will actually call out to user code (the user method handler), so even the internal things are actually external.  I would have said that they are also initiated by the library and not under a lock, but we can envision a particularly insidious example using an in-process transport where both sides of the RPC share the same lock and that the server method handler grabs the lock as part of its work and then initiates a client RPC. Does removing the internal option destroy performance?

Reply:
I can give it another shot without the internal thing and check performance.

Reply:
Removing the "internalization" results in a loss of performance.

Synthetic test with this PR but without "internalization" -->
Final Score (geometric mean): 0.902

Synthetic test with this PR but with "internalization" -->
Final Score (geometric mean): 0.974

So, I am going to try and go ahead with internalization.